clang/include/clang/Basic/DiagnosticSemaKinds.td

```diff
@@ -1637,10 +1637,8 @@ def err_inline_namespace_std : Error<
 def err_unexpected_friend : Error<
   "friends can only be classes or functions">;
 def ext_enum_friend : ExtWarn<
-  "befriending enumeration type %0 is a C++11 extension">, InGroup<CXX11>;
-def warn_cxx98_compat_enum_friend : Warning<
-  "befriending enumeration type %0 is incompatible with C++98">,
-  InGroup<CXX98Compat>, DefaultIgnore;
+  "elaborated enumeration type cannot be a friend">,
```

Comment:
I don't think elaborated enumeration type is a thing (although elaborated type specifier is). I think the simplest solution is to remove "elaborated" from diagnostic message.

Reply:
@Endilll I tried to follow the existing "naming convention" used in other diagnostics to strike a balance between being correct & comprehensible for users. Removing "elaborated" isn't quite correct because an enumeration type _can_ be a friend (e.g. `enum class E; struct A { friend E; };`; such a declaration is ignored) -- you just can't declare a friend using an elaborated-type-specifier starting with `enum` (i.e. `friend enum E`). 

Here is an alternative if you think changing this is necessary:
> A friend declaration cannot use an elaborated type specifier starting with `enum`

Reply:
I think now I see what do you want to achieve. I definitely missed that enums can be befriended, just not using elaborated type specifier syntax. Sorry for some of the comments I left coming from a wrong perspective.

What if we skip complicated standardese altogether?
> 'enum' keyword is not allowed after `friend` keyword

Possibly followed by
> Omit 'enum' if you intend to befriend an enumeration

Reply:
> What if we skip complicated standardese altogether?
> 
> > 'enum' keyword is not allowed after 'friend' keyword

This isn't quite correct either because `enum` _can_ follow the `friend` keyword 😛, e.g. 
```cpp
enum class E;
struct A
{
    friend enum E f(); // ok
};
```

> Possibly followed by
> 
> > Omit 'enum' keyword if you intend to befriend an enumeration

I don't think a note would be particularly helpful since the friend declaration would be ignored anyways...

Reply:
> This isn't quite correct either because enum can follow the friend keyword 😛, e.g.

Ugh, you are right. I think it even defeats your initial suggestions, which read `A friend declaration cannot use an elaborated type specifier starting with enum`, because it can when befriending a function, as you just showed. I guess no way around standardese then: `Elaborated type specifier cannot be used to befriend an enumeration. Consider omitting 'enum' keyword`. I'd like the wording to be accurate on terminology, yet actionable for users who are not well-versed in standardese.

Reply:
> I don't think a note would be particularly helpful since the friend declaration would be ignored anyways...

Would it? I'd expect enumerators to have access to non-public constants.
Even if doesn't hold, we should leave the opinion of usefulness to more opinionated tools like clang-tidy.

Reply:
@Endilll 
> Would it? I'd expect enumerators to have access to non-public constants.

The enumerators do not get any special access. Per [[class.friend] p3](http://eel.is/c++draft/class.friend#3):
> If the type specifier in a friend declaration designates a (possibly cv-qualified) class type, that class is declared as a friend; otherwise, the friend declaration is ignored.

Reply:
@Endilll I think I will settle for "A friend enumeration type cannot be declared using an elaborated type specifier" or "An elaborated type specifier starting with 'enum' cannot be declared as a friend". Which do you prefer, or should we leave the current wording for the time being?

Reply:
> ISO C++ does not permits this declaration to be declared as friend because 'enum E' is an elaborated specifier.

Bonus point for a fix-it removing enum

Reply:
I can't say I particularly like any of the three ideas suggested:
> A friend enumeration type cannot be declared using an elaborated type specifier
> An elaborated type specifier starting with 'enum' cannot be declared as a friend

I find both not too actionable for people out of the loop on the Standard. Among those two options, I prefer the latter.

> ISO C++ does not permits this declaration to be declared as friend because 'enum E' is an elaborated specifier.

This is somewhat better, but isn't there an implication that elaborated type specifiers are not permitted in friend declarations?

> should we leave the current wording for the time being?

Sorry, I don't find the current wording `elaborated enumeration type cannot be a friend` acceptable.

Reply:
I pinged @AaronBallman offline. He promised to take a look at this.

Reply:
> ISO C++ does not permits this declaration to be declared as friend because 'enum E' is an elaborated enum specifier.

There. Something along these lines

Reply:
Another alternative: "friend declared using elaborated type specifier must be a class"

Reply:
I'd go with: 

`friend declaration of %0 with an elaborated specifier is not allowed in ISO C++`

and I agree that a fix-it to remove the `enum` keyword would be great if we can do it (not all users are going to know what an "elaborated specifier" is, so the fix-it gives a bit of context to the diagnostic).

Reply:
> > ISO C++ does not permits this declaration to be declared as friend because 'enum E' is an elaborated enum specifier.
> 
> There. Something along these lines

Not a huge fan of this phrasing. The wording of the diagnostic should already imply that such a declaration is invalid. Moreover:
- "elaborated enum specifier" isn't any "more correct" than "elaborated enum type"
- "declaration to be declared" is a bit redundant
- the phrasing implies that the declaration might be well-formed when written without `friend` (which it isn't).

Reply:
Added a fixit. Since everyone seems to be learning towards "elaborated enum specifier", would everyone be happy with "elaborated enum specifier cannot be declared as friend"?

Reply:
I'd be okay with that, but I'd say `cannot be declared as a friend` just so it reads slightly better.